language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/MappingContextProvider.java

```diff
@@ -0,0 +1,189 @@
+import io.ballerina.compiler.api.symbols.SymbolKind;
+import io.ballerina.compiler.api.symbols.TypeDescKind;
+import io.ballerina.compiler.api.symbols.TypeSymbol;
+import io.ballerina.compiler.api.symbols.UnionTypeSymbol;
+import io.ballerina.compiler.api.symbols.VariableSymbol;
+import io.ballerina.compiler.syntax.tree.ComputedNameFieldNode;
+import io.ballerina.compiler.syntax.tree.Node;
+import io.ballerina.compiler.syntax.tree.NonTerminalNode;
+import io.ballerina.compiler.syntax.tree.QualifiedNameReferenceNode;
+import io.ballerina.compiler.syntax.tree.SpecificFieldNode;
+import io.ballerina.compiler.syntax.tree.SyntaxKind;
+import io.ballerina.compiler.syntax.tree.Token;
+import org.apache.commons.lang3.tuple.Pair;
+import org.ballerinalang.langserver.common.utils.CommonUtil;
+import org.ballerinalang.langserver.common.utils.completion.QNameReferenceUtil;
+import org.ballerinalang.langserver.commons.BallerinaCompletionContext;
+import org.ballerinalang.langserver.commons.completion.LSCompletionItem;
+import org.ballerinalang.langserver.completions.SnippetCompletionItem;
+import org.ballerinalang.langserver.completions.SymbolCompletionItem;
+import org.ballerinalang.langserver.completions.builder.VariableCompletionItemBuilder;
+import org.ballerinalang.langserver.completions.providers.AbstractCompletionProvider;
+import org.ballerinalang.langserver.completions.util.ContextTypeResolver;
+import org.ballerinalang.langserver.completions.util.Snippet;
+import org.eclipse.lsp4j.CompletionItem;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+/**
+ * Abstract completion provider for mapping constructor context.
+ *
+ * @param <T> mapping constructor node type.
+ * @since 2.0.0
+ */
+public abstract class MappingContextProvider<T extends Node> extends AbstractCompletionProvider<T> {
+
+    protected abstract Map<String, RecordFieldSymbol> getValidFields(T node,
+                                                                     RecordTypeSymbol recordTypeSymbol);
+
+    public MappingContextProvider(Class<T> attachmentPoint) {
+        super(attachmentPoint);
+    }
+
+    protected Predicate<Symbol> getVariableFilter() {
+        return CommonUtil.getVariableFilterPredicate().or(symbol -> symbol.kind() == SymbolKind.CONSTANT);
+    }
+
+    protected boolean withinValueExpression(BallerinaCompletionContext context, NonTerminalNode evalNodeAtCursor) {
+        Token colon = null;
+
+        if (evalNodeAtCursor.kind() == SyntaxKind.SPECIFIC_FIELD) {
+            colon = ((SpecificFieldNode) evalNodeAtCursor).colon().orElse(null);
+        } else if (evalNodeAtCursor.kind() == SyntaxKind.COMPUTED_NAME_FIELD) {
+            colon = ((ComputedNameFieldNode) evalNodeAtCursor).colonToken();
+        }
+
+        if (colon == null) {
+            return false;
+        }
```

Comment:
just return at the else clause of the previous if-else ladder

Reply:
Done